kv/memdb/memdb.go

```diff
@@ -21,12 +21,42 @@ import (
 	"github.com/pingcap/tidb/util/fastrand"
 )
 
+// KeyFlags are metadata associated with key
+type KeyFlags uint8
+
 const (
+	flagNeedLocked KeyFlags = 1 << iota
+
+	// flagTouched is a internal flag to help `Mark` operation.
+	// The default value of `KeyFlags` is 0 and the flags returned by any operation will set `flagTouched`.
+	// When merge two flags, if the newer one haven't set the `flagTouched`, the newer value will be ignored.
+	flagTouched KeyFlags = 0x80
+
 	maxHeight      = 16
 	nodeHeaderSize = int(unsafe.Sizeof(nodeHeader{}))
 	initBlockSize  = 4 * 1024
 )
 
+// HasNeedLocked returns whether the associated key required lock.
+func (m KeyFlags) HasNeedLocked() bool {
+	return m&flagNeedLocked != 0
+}
+
+func (m KeyFlags) isTouched() bool {
+	return m&flagTouched != 0
+}
+
+// Merge used to merge two KeyFlags.
+func (m KeyFlags) Merge(old KeyFlags) KeyFlags {
+	// Only consider flagPresumeKeyNotExists in merge operation for now.
```

Comment:
This strange policy only valid when we just considering `flagPresumeKeyNotExists`.

In this case, we do not need `isTouched` and `Merge` because we will never **unmark a flag** manually. We can use a naive **bit logical or** to replace the `Merge` policy.